src/backend/distributed/transaction/shard_transaction.c

```diff
@@ -0,0 +1,212 @@
+
+		if (connection == NULL)
+		{
+			List *abortConnectionList = ConnectionList(shardConnectionHash);
+			CloseConnections(abortConnectionList);
+
+			ereport(ERROR, (errmsg("could not establish a connection to all "
+								   "placements of shard %lu", shardId)));
+		}
+
+		transactionConnection = palloc0(sizeof(TransactionConnection));
+
+		transactionConnection->connectionId = shardConnections->shardId;
+		transactionConnection->transactionState = TRANSACTION_STATE_INVALID;
+		transactionConnection->connection = connection;
+
+		connectionList = lappend(connectionList, transactionConnection);
+	}
+
+	shardConnections->connectionList = connectionList;
+}
+
+
+/*
+ * ConnectionList flattens the connection hash to a list of placement connections.
+ */
+List *
+ConnectionList(HTAB *connectionHash)
+{
+	List *connectionList = NIL;
+	HASH_SEQ_STATUS status;
+	ShardConnections *shardConnections = NULL;
+
+	hash_seq_init(&status, connectionHash);
+
+	shardConnections = (ShardConnections *) hash_seq_search(&status);
+	while (shardConnections != NULL)
+	{
+		List *shardConnectionsList = list_copy(shardConnections->connectionList);
+		connectionList = list_concat(connectionList, shardConnectionsList);
+
+		shardConnections = (ShardConnections *) hash_seq_search(&status);
+	}
+
+	return connectionList;
+}
+
+
+/*
+ * CloseConnections closes all connections in connectionList.
+ */
+void
+CloseConnections(List *connectionList)
+{
+	ListCell *connectionCell = NULL;
+
+	foreach(connectionCell, connectionList)
+	{
+		TransactionConnection *transactionConnection =
+			(TransactionConnection *) lfirst(connectionCell);
+		PGconn *connection = transactionConnection->connection;
+
+		PQfinish(connection);
+	}
+}
+
+
+/*
+ * OpenTransactionsToAllShardPlacements opens connections to all placements of
+ * the given shard Id Pointer List and returns the hash table containing the connections.
+ * The resulting hash table maps shardIds to ShardConnection structs.
+ */
+HTAB *
+OpenTransactionsToAllShardPlacements(List *shardIdList, char *relationOwner)
+{
+	HTAB *shardConnectionHash = CreateShardConnectionHash();
+	ListCell *shardCell = NULL;
+	ListCell *connectionCell = NULL;
+	List *connectionList = NIL;
+
+	foreach(shardCell, shardIdList)
+	{
+		uint64 *shardIdPointer = (uint64 *) lfirst(shardCell);
+		uint64 shardId = (*shardIdPointer);
+
+		OpenConnectionsToShardPlacements(shardId, shardConnectionHash, relationOwner);
+	}
+	
+	connectionList = ConnectionList(shardConnectionHash);
+	
+	foreach(connectionCell, connectionList)
+	{
+		TransactionConnection *transactionConnection =
+			(TransactionConnection *) lfirst(connectionCell);
+		PGconn *connection = transactionConnection->connection;
+		PGresult *result = NULL;
+		
+		result = PQexec(connection, "BEGIN");
+		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		{
+			WarnRemoteError(connection, result);
+			ereport(ERROR, (errmsg("could not send ddl command to shard placement")));
```

Comment:
Maybe just ReraiseRemoteError. If this function has another caller, the error message doesn't make much sense.